networkdb/delegate.go

```diff
@@ -16,92 +16,61 @@ func (d *delegate) NodeMeta(limit int) []byte {
 	return []byte{}
 }
 
-// getNode searches the node inside the tables
-// returns true if the node was respectively in the active list, explicit node leave list or failed list
-func (nDB *NetworkDB) getNode(nEvent *NodeEvent, extract bool) (bool, bool, bool, *node) {
-	var active bool
-	var left bool
-	var failed bool
-
-	for _, nodes := range []map[string]*node{
-		nDB.failedNodes,
-		nDB.leftNodes,
-		nDB.nodes,
-	} {
-		if n, ok := nodes[nEvent.NodeName]; ok {
-			active = &nodes == &nDB.nodes
-			left = &nodes == &nDB.leftNodes
-			failed = &nodes == &nDB.failedNodes
-			if n.ltime >= nEvent.LTime {
-				return active, left, failed, nil
-			}
-			if extract {
-				delete(nodes, n.Name)
-			}
-			return active, left, failed, n
-		}
-	}
-	return active, left, failed, nil
-}
-
 func (nDB *NetworkDB) handleNodeEvent(nEvent *NodeEvent) bool {
 	// Update our local clock if the received messages has newer
 	// time.
 	nDB.networkClock.Witness(nEvent.LTime)
 
 	nDB.RLock()
-	active, left, _, n := nDB.getNode(nEvent, false)
+	defer nDB.RUnlock()
```

Comment:
Do you really want to hold on the nDB lock for the entire function call ?

Reply:
yep, the function is fairly short and avoids dealing with crazy race where other node events can come in the middle and change the state. If an event come and is fresh, has to be atomically handled in one shot. In the previous logic every time you were taking back the log you had to start all the checks from scratch because another event maybe newer changed the node state. At the end the increase of complexity and exposure to race condition in this part of the code is not worth the time of the lock. These are also node events so should be not super frequent as table events, also if synced every 30s by push/pull, the current logic will anyway return at line 35